routes/artifacts.js

```diff
@@ -0,0 +1,597 @@
+      bucket = this.publicBucket;
+    }
+    if (artifact.details.bucket === this.privateBucket.bucket) {
+      bucket = this.privateBucket;
+    }
+    // Create put URL
+    var putUrl = await this.privateBucket.createPutUrl(
+      artifact.details.prefix, {
+      contentType:      artifact.contentType,
+      expires:          30 * 60 + 10 // Add 10 sec for clock drift
+    });
+    return res.reply({
+      storageType:  's3',
+      contentType:  artifact.contentType,
+      expires:      expiry.toJSON(),
+      putUrl:       putUrl
+    });
+  }
+
+  // Reply with SAS for azure
+  if (artifact.storageType === 'azure') {
+    var expiry = new Date(new Date().getTime() + 30 * 60 * 1000);
+    // Generate SAS
+    var putUrl = this.blobStore.generateWriteSAS(
+      artifact.details.path, {
+      expiry:         expiry
+    });
+    return res.reply({
+      storageType:  'azure',
+      contentType:  artifact.contentType,
+      expires:      expiry.toJSON(),
+      putUrl:       putUrl
+    });
+  }
+
+  // For 'reference' and 'error' the response is simple
+  return res.reply({
+    storageType:    storageType
+  });
+});
+
+/** Reply to an artifact request using taskId, runId, name and context */
+var replyWithArtifact = async function(taskId, runId, name, res) {
+  // Load artifact meta-data from table storage
+  var artifact = await this.Artifact.load({
+    taskId:     taskId,
+    runId:      runId,
+    name:       name
+  }, true);
+
+  // Give a 404, if the artifact couldn't be loaded
+  if (!artifact) {
+    return res.status(404).json({
+      message:  "Artifact not found"
+    });
+  }
+
+  // Handle S3 artifacts
+  if(artifact.storageType === 's3') {
+    // Find url
+    var url = null;
+    var prefix = artifact.details.prefix;
+    if (artifact.details.bucket === this.publicBucket.bucket) {
+      url = this.publicBucket.createGetUrl(prefix);
+    }
+    if (artifact.details.bucket === this.privateBucket.bucket) {
+      url = await this.privateBucket.createSignedGetUrl(prefix, {
+        expires:    30 * 60
+      });
+    }
+    assert(url, "Url should have been constructed!");
+    return res.redirect(303, url);
+  }
+
+  // Handle azure artifacts
+  if(artifact.storageType === 'azure') {
+    if (artifact.details.container !== this.blobStore.container) {
+      debug("[alert-operator], Unknown container: %s, for artifact: %j",
+            artifact.details.container, artifact.json());
+    }
+    // Generate URL expiration time
+    var expiry = new Date();
+    expiry.setMinutes(expiry.getMinutes() + 30);
+    // Create and redirect to signed URL
+    return res.redirect(303, this.blobStore.createSignedGetUrl(
+      artifact.details.path, {
+      expiry:   expiry
+    }));
+  }
+
+  // Handle redirect artifacts
+  if (artifact.storageType === 'reference') {
+    return res.redirect(303, artifact.details.url);
+  }
+
+  // Handle error artifacts
+  if (artifact.storageType === 'error') {
```

Comment:
I wonder if there's value to creating a set of objects, S3, Azure, Reference and Error which each know how to do each operation and set a variable once in the beginning of the function that says which object to call.  It seems like a great place for bugs to slip in where we have these repeating branches